Update documentation for wide character support

avh4 opened the pull request:
The guide says "Brick relies on Vty's use of the [utf8proc](http://julialang.org/utf8proc/) library to determine the column width of each character rendered." but use of utf8proc was removed in vty-5.13  https://github.com/jtdaugherty/vty/commit/cd534c1028b3efde9c13218a10b7e085958496bc.

This updates the brick guide to point to how to configure the new way of computing character widths that was since added.  (It took me a few hours of digging through issue histories to find this and understand that it was the current correct way!)

avh4 commented:
For reference, past related issues:
- https://github.com/jtdaugherty/brick/issues/265
- https://github.com/jtdaugherty/brick/issues/356
- https://github.com/jtdaugherty/vty/issues/175 (implements the current solution)

xsebek commented:
Interesting, should applications using brick and wide characters suggest users create the configuration for their terminal?

Or would it be best to give an option like `--unicode iterm2` and have the brick application generate it for the user? :confused: 

The Matterhorn application mentioned in one of those issues does not suggest a solution in its FAQ:
- https://github.com/matterhorn-chat/matterhorn/blob/master/docs/FAQ.md#does-matterhorn-support-graphical-emoji

jtdaugherty commented:
Unfortunately it's a little worse than the current state. When I added the "width table" feature to Vty, it seemed at the time that it could result in width tables that were on average more likely to be correct than the width table that was built into the library at the time. That turned out to be quite wrong; the process of querying the terminal got wrong answers sometimes, mostly because the character ranges that get queried include combining characters. So the width table approach itself is actually not going to work in the long run because a proper width calculation approach needs to perform Unicode-aware lookahead. This is because there are lots of Unicode character sequences that need to be combined *before* their width can be determined.

So at this point it's not quite right to say that the table generation approach is the best way to go; we provided it in case someone found that, in their situation, it gave a better answer than the built-in (old) table.

With that said, I'm torn on what to do about it because I can't recommend that people always use the table generator, but clearly the old built-in table will be wrong for wide characters (or combinable ones) added in new revisions of Unicode. We also don't know in general how smart the terminal's Unicode width calculations are, and we've found that in some cases they have just as naive an approach as Vty has without the tables set up (i.e. a fixed table with no lookahead).

My recommendation at this point is to avoid wide characters because in my view, wide character support in terminals in general is just a huge mess. We've tried hacks in Vty to compensate, but they break as much as they fix because different terminals are at different levels of broken, and those brokenness levels change over time as Unicode is updated!

As far as this patch is concerned, I don't want Brick's documentation to say anything about the Vty generated table configuration quite yet because (as I said above) I wouldn't recommend that be the first approach people take. I'd be okay with the Brick docs just saying that Brick depends on Vty's *API* for wide characters. Under the hood that API could use different mechanisms independently.

Thanks!

avh4 commented:
So the situation I'm trying to improve is the one I was in:
- I started using brick
- eventually I tried putting a few emojis in the UI (in my case it was 😵 and 🌘)
- the layout broke from the emoji
- I searched for issues and documentation and eventually read that things were supposed to work
- I confirmed I had the latest brick and vty versions
- things still didn't work, so I had to re-read all the existing issues -- but it's quite confusing because you eventually realized that there were different fixes at different times and you have to start looking at the timestamps of the comments to understand the sequence of events and the current state of things
- I even had to figure out how to write a little c program to determine whether or not utf8proc was working correctly because I thought it was meant to be being used :sweat_smile:
- I also tried the workaround mentioned in https://github.com/jtdaugherty/brick/issues/356#issuecomment-1001247381, which seemed to work at first, but later I found it caused bugs with screen redrawing

Anyway, as soon as I ran `vty-build-table-width -u`, everything worked fine for me (so far, at least...).  And I'm also planning to have my app use the table-building API on first startup before I release it.

So I'd like to propose at least one of the following:
- I could modify this PR to still link there, but to only point you there if you run into issues, and not as a general recommendation (or maybe even add your quote saying you officially don't  recommend it)
- cancel this PR, and I could just add a note as a comment to the past issues I found when looking into this
- slightly modify the default behavior so that emoji default to 2 width instead of 1 width (like 😵 and 🌘)

Would the first of those (link but with a warning) be okay?

jtdaugherty commented:
@avh4 thanks for such a detailed explanation of how you got to this point. It's both helpful for me to hear it, and saddening to hear that it was way more indirect than I'd like. Since it's also no fun for me to apologize for the situation several times a year (you aren't the first person to encounter these issues), it would also help me to set some expectations about this more clearly in the documentation. It seems like an FAQ entry as well as some User Guide material on this would have been helpful. I'm happy to make some updates along those lines.

As for this patch, your first option works for me. Thank you!

avh4 commented:
Okay, I updated it to hopefully make the current recommendation clear, and also added it to the FAQ 8baa49f29b4f92e3211e754386e2764ef378de32

How does that look?

jtdaugherty commented:
Looks good to me. Thank you!
